arroyo/processing/strategies/collect.py

```diff
@@ -21,7 +32,210 @@ class OffsetRange:
+    """
+    Represents a batch of messages with the range of offsets covered.
+    The range of offsets is represented as a mapping between partitions
+    and offset ranges.
+    The lowest offset in each offset range is inclusize. The highest is
+    esclusive.
+    """
+
+    messages: Sequence[Message[TPayload]]
+    offsets: Mapping[Partition, OffsetRange]
+    duration: float
+
+    def __len__(self) -> int:
+        return len(self.messages)
+
+    def __repr__(self) -> str:
+        return f"<{type(self).__name__}: {len(self)} message{'s' if len(self) != 1 else ''}, open for {self.duration:0.2f} seconds>"
+
+
+class BatchBuilder(Generic[TPayload]):
+    def __init__(
+        self,
+        max_batch_time: float,
+        max_batch_size: int,
+    ) -> None:
+        self.__max_batch_time = max_batch_time
+        self.__max_batch_size = max_batch_size
+        self.__messages: MutableSequence[Message[TPayload]] = []
+        self.__offsets: MutableMapping[Partition, OffsetRange] = {}
+        self.__init_time = time.time()
+
+    def is_ready(self) -> bool:
+        return len(self.__messages) >= self.__max_batch_size or (
+            time.time() > self.__init_time + self.__max_batch_time
+        )
+
+    def append(self, message: Message[TPayload]) -> None:
+        if message.partition in self.__offsets:
+            self.__offsets[message.partition].hi = message.next_offset
+            self.__offsets[message.partition].timestamp = message.timestamp
+        else:
+            self.__offsets[message.partition] = OffsetRange(
+                message.offset, message.next_offset, message.timestamp
+            )
+
+        self.__messages.append(message)
+
+    def flush(self) -> MessageBatch[TPayload]:
+        return MessageBatch(
+            messages=self.__messages,
+            offsets=self.__offsets,
+            duration=time.time() - self.__init_time,
+        )
+
+
+class BatchStep(ProcessingStep[TPayload]):
+    """
+    Accumulates messages into a batch. When the batch is full this
+    strategy submits it to the next step.
+
+    A batch is represented as a `MessageBatch` object, which includes
+    both the messages and the offset ranges.
+
+    A message is closed and submitted when the maximum number of messages
+    is received or when the deadline has passed.
+    """
+
+    def __init__(
+        self,
+        max_batch_time: float,
+        max_batch_size: int,
+        next_step: ProcessingStep[MessageBatch[TPayload]],
+    ) -> None:
+        self.__max_batch_time = max_batch_time
+        self.__max_batch_size = max_batch_size
+        self.__next_step = next_step
+        self.__batch_builder: Optional[BatchBuilder[TPayload]] = None
+        self.__closed = False
+
+    def __flush(self, robust: bool) -> None:
+        assert self.__batch_builder is not None
+        batch = self.__batch_builder.flush()
+        try:
+            if len(batch) > 0:
+                last_msg = batch.messages[-1]
```

Comment:
this seems like something you can keep track of in the `MessageBatch`

Reply:
Not sure I follow. What would you keep track of in `MessageBatch` ?

Reply:
the last message

Reply:
done